### PR TITLE
Tree: Path prefix

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Add an optional prefix to the cursor path methods.

This makes composing cursors easier. For example BasicChunkCursor will want to delegate to UniformChunkCursor for the leaves of trees, and needs a way for the paths to compose correctly.

This also helps clear up the semantics of the path getting methods for non-rooted cases, and provides some functionality that might be generally useful.

Where possible, cursors implement this functionality directly to avoid the overhead of wrapping the cursor.
For other cases, a utility is provided to prefix any cursor.

## Breaking Changes

As the tree package is not exported, this does not count as a published breaking change, however is is breaking at the package API level in two respects:
1. External implementers of cursor become incorrect (but still build). Currently there should not be any, so this is fine.
2. Anyone calling getFieldPath or getPath with an argument will incorrectly pass that as the prefix instead of having it discarded. This is unlikely to happen with methods but is possible (ex: passing the result of getPath.bind() to array.map will no longer work)

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: packages/dds/tree/src/test/cursorTestSuite.ts [hunk 1]
```diff
@@ -566,13 +568,93 @@ function testTreeCursor<TData, TCursor extends ITreeCursor>(config: {
...
+                    expectEqualPaths(cursor.getPath(undefined), parent);
+                    expectEqualPaths(cursor.getPath({}), parent);
+
+                    const prefixParent: UpPath = {
+                        parent: undefined,
+                        parentField: brand("prefixParentField"),
+                        parentIndex: 5,
+                    };
+                    if (extraRoot) {
+                        // TODO
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/cursorTestSuite.ts:601]: I need to fix this test coverage gap, and see if this is why BasiChunkCursor is passing tests (its missing the prefix logic, and should fail)

File: packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            );
+            assert(
+                compareUpPaths(prefixPath({ rootFieldOverride: brand("y") }, root), {
+                    parent: undefined,
+                    parentField: brand("y"),
+                    parentIndex: 5,
+                }),
+            );
+            assert(
+                compareUpPaths(prefixPath({ parent: root }, root), {
```
Comment (jenn-le) [inline packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts:52]: This case is confusing to me. What does making the root path's parent be itself actually mean?

File: packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        it("wrapped root", () => {
+            assert(
+                compareUpPaths(prefixPath({ indexOffset: 1 }, root), {
+                    parent: undefined,
+                    parentField: brand("x"),
+                    parentIndex: 6,
+                }),
+            );
+            assert(
+                compareUpPaths(prefixPath({ rootFieldOverride: brand("y") }, root), {
```
Comment (jenn-le) [inline packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts:45]: I'm still not sure I understand the purpose of the rootFieldOverride. So if a cursor wraps another cursor using this prefix, does it just mean that the outer cursor moves to the same index but in the y field? And so, is it just a quick way to shift around fields and if so, why do we only allow an override at the root?

File: packages/dds/tree/src/core/tree/cursor.ts [hunk 4]
```diff
@@ -211,6 +209,33 @@ export interface ITreeCursor {
     readonly value: Value;
 }
 
+/**
+ * Prefix to use as the root of a path.
+ *
+ * This describes a location that the content accessible through the cursor could be parented,
+ * and is used to adjust the paths to be as if the content were parented there.
+ *
+ * Typically this is used to make trees appear to be part of a forest when getting their path.
```
[the remark this answers is not in the document]
Comment (jenn-le) [reply packages/dds/tree/src/core/tree/cursor.ts:218]: Oh, does this mean to make them appear as though they're not detached?

File: packages/dds/tree/src/core/tree/cursor.ts [hunk 5]
```diff
@@ -100,29 +100,27 @@ export interface ITreeCursor {
      * @returns a path to the current field. See {@link FieldUpPath}.
      *
      * Only valid when `mode` is `Fields`.
-     * Assumes this cursor has a root node where its field keys are actually detached sequences.
-     * If the cursor is not rooted at such a node,
-     * calling this function is invalid, and the returned FieldUpPath (if any) may not be meaningful.
+     *
+     * If no prefix is provided, assumes this cursor is treated as if it has a root node where its field keys are actually detached sequences.
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/core/tree/cursor.ts:104]: I have rewritten these docs. Maybe its better now.

File: packages/dds/tree/src/core/tree/cursor.ts [hunk 6]
```diff
@@ -100,29 +100,27 @@ export interface ITreeCursor {
...
 
     /**
      * @returns a path to the current node. See {@link UpPath}.
      *
      * Only valid when `mode` is `Nodes`.
-     * Assumes this cursor has a root node where its field keys are actually detached sequences.
-     * If the cursor is not rooted at such a node,
-     * calling this function is invalid, and the returned UpPath (if any) may not be meaningful.
+     *
+     * If no prefix is provided, assumes this cursor is treated as if it has a root node where its field keys are actually detached sequences.
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/core/tree/cursor.ts:124]: I have rewritten these docs. Maybe its better now.

File: packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts:45]: I have updated the docs on PathRootPrefix. Does it address this well enough now?

File: packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts:52]: Like the two cases above, this is just resting one of the fields in PathRootPrefix, in this case `parent`. `root` is simply the simplest UpPath thats in scope to use as the parent, and also the simplest UpPath we can apply the prefix to.

File: packages/dds/tree/src/core/tree/cursor.ts [hunk 7]
```diff
@@ -211,6 +221,36 @@ export interface ITreeCursor {
     readonly value: Value;
 }
 
+/**
+ * Prefix to apply to the as the root of a {@link UpPath} or {@link FieldUpPath}.
```
Comment (jenn-le) [inline packages/dds/tree/src/core/tree/cursor.ts:225]:
```suggestion
 * Prefix to apply as the root of a {@link UpPath} or {@link FieldUpPath}.
```

File: packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts (hunk 3 above)
Comment (jenn-le) [reply packages/dds/tree/src/test/feature-libraries/treeCursorUtils.spec.ts:45]: Yes, that helps a lot. Thank you!

**Review by jenn-le — APPROVED** [review]

**Pull request merged**